src/platform/ASR/PlatformManagerImpl.cpp

```diff
@@ -0,0 +1,353 @@
+            return CHIP_ERROR_INTERNAL;
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+#else
+CHIP_ERROR PlatformManagerImpl::InitLwIPCoreLock(void)
+{
+    return Internal::InitLwIPCoreLock();
+}
+#endif
+
+void PlatformManagerImpl::_Shutdown()
+{
+    uint64_t upTime = 0;
+
+    if (GetDiagnosticDataProvider().GetUpTime(upTime) == CHIP_NO_ERROR)
+    {
+        uint32_t totalOperationalHours = 0;
+
+        if (ConfigurationMgr().GetTotalOperationalHours(totalOperationalHours) == CHIP_NO_ERROR)
+        {
+            ConfigurationMgr().StoreTotalOperationalHours(totalOperationalHours + static_cast<uint32_t>(upTime / 3600));
+        }
+        else
+        {
+            ChipLogError(DeviceLayer, "Failed to get total operational hours of the Node");
+        }
+    }
+    else
+    {
+        ChipLogError(DeviceLayer, "Failed to get current uptime since the Node’s last reboot");
+    }
+#if CONFIG_ENABLE_ASR_LEGA_RTOS
+    //
+    // Call up to the base class _Shutdown() to perform the actual stack de-initialization
+    // and clean-up
+    //
+
+    (void) _StopEventLoopTask();
+
+    // the task thread is self terminating, we might have to wait if it's still processing
+    while (true)
+    {
+        if (kTaskRunningEventFlag == 0)
+        {
+            break;
+        }
+        lega_rtos_delay_milliseconds(1);
+    }
+
+    ChipLogError(DeviceLayer, "StopEventLoopTask done.");
+
+    lega_rtos_deinit_event_flags(&mEventFlags);
+    lega_rtos_deinit_queue(&mEventQueue);
+    lega_rtos_deinit_timer(&mTimer);
+    lega_rtos_deinit_mutex(&mChipMutex);
+    lega_rtos_deinit_mutex(&mEventMutex);
+
+    Internal::GenericPlatformManagerImpl<PlatformManagerImpl>::_Shutdown();
+#else
+    Internal::GenericPlatformManagerImpl_FreeRTOS<PlatformManagerImpl>::_Shutdown();
+#endif
+}
+
+#if CONFIG_ENABLE_ASR_LEGA_RTOS
+void PlatformManagerImpl::SetEventFlags(uint32_t flags)
+{
+    if (lega_rtos_set_event_flags(&mEventFlags, flags) != kNoErr)
+    {
+        ChipLogError(DeviceLayer, "%s lega_rtos_set_event_flags %08lx", __func__, flags);
+    }
+}
+
+void PlatformManagerImpl::HandleTimerEvent(void)
+{
+    const CHIP_ERROR err = static_cast<System::LayerImplFreeRTOS &>(DeviceLayer::SystemLayer()).HandlePlatformTimer();
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(DeviceLayer, "HandlePlatformTimer %ld", err.AsInteger());
+    }
+}
+
+void PlatformManagerImpl::HandlePostEvent(void)
+{
+    OSStatus result;
+    ChipDeviceEvent event;
+
+    /* Check the event queue. */
+    if (lega_rtos_is_queue_empty(&mEventQueue))
+    {
+        return;
+    }
+
+    /* Pop one event from the event queue. */
+    result = lega_rtos_pop_from_queue(&mEventQueue, &event, LEGA_WAIT_FOREVER);
+
+    if (kNoErr != result)
+    {
+        ChipLogError(DeviceLayer, "lega_rtos_pop_from_queue %u", result);
+        return;
+    }
+
+    /* Process this event. */
+    DispatchEvent(&event);
```

Comment:
Where are we locking mChipMutex while this happens?  For that matter, is the queue access ok if it's happening from multiple threads at once?  @andy31415 @tx2rx

Reply:
@bzbarsky-apple  we use EventFlags to resolve this , only one event can pop from the event queue, and the thread(EventLoopTaskMain) only create once.

Reply:
@tx2rx yes, but what happens when _other threads_ try to LockChipStack and then touch things, without going through the event loop at all?

Reply:
in addition, src/include/platform/internal/GenericPlatformManagerImpl_FreeRTOS.ipp may cause the same problem

Reply:
No, that one does not have the problem: it holds the lock while dispatching events.

Reply:
resolved at https://github.com/project-chip/connectedhomeip/pull/26645